tendermint/src/lite/public.rs

```diff
@@ -0,0 +1,170 @@
+use crate::block::Height;
+use crate::lite::*;
+use std::time::{Duration, SystemTime};
+
+// Verify a single untrusted header against a trusted state.
+// Includes all validation and signature verification.
+// Not publicly exposed since it does not check for expiry
+// and hence it's possible to use it incorrectly.
+// If trusted_state is not expired and this returns Ok, the
+// untrusted_sh and untrusted_next_vals can be considered trusted.
+fn verify_single<TS, SH, C, L>(
+    trusted_state: &TS,
+    untrusted_sh: &SH,
+    untrusted_vals: &C::ValidatorSet,
+    untrusted_next_vals: &C::ValidatorSet,
+    trust_threshold: &L,
+) -> Result<(), Error>
+where
+    TS: TrustedState<LastHeader = SH, ValidatorSet = C::ValidatorSet>,
+    SH: SignedHeader<Commit = C>,
+    C: Commit,
+    L: TrustThreshold,
+{
+    // ensure the new height is higher
+    let untrusted_height = untrusted_sh.header().height();
+    let trusted_height = trusted_state.last_header().header().height();
+    if untrusted_height <= trusted_height {
+        // TODO: return err
+    };
+
+    // validate the untrusted header against its commit, vals, and next_vals
+    let untrusted_header = untrusted_sh.header();
+    let untrusted_commit = untrusted_sh.commit();
+    validate_vals_and_commit(untrusted_header, untrusted_commit, untrusted_vals)?;
+    validate_next_vals(untrusted_header, untrusted_next_vals)?;
+
+    // if the new height is not sequential, check if we can skip
+    if untrusted_height > trusted_height.increment() {
+        let trusted_vals = trusted_state.validators();
+        verify_commit_trusting(trusted_vals, untrusted_commit, trust_threshold)?;
```

Comment:
In case of sequential headers we need to check also if 'if trusted_header.NextValidatorsHash != untrusted_header.ValidatorsHash'.

Reply:
Good catch!

Reply:
fixed in 39be6313e4861ecc836ebe47e8e11a2512f88115